saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -167,93 +166,113 @@ public boolean supports(Class<?> authentication) {
 											String recipient,
-											Response samlResponse) throws AuthenticationException {
+											Response samlResponse) throws Saml2AuthenticationException {
+		//optional validation if the response contains a destination
 		if (hasText(samlResponse.getDestination()) && !recipient.equals(samlResponse.getDestination())) {
-			throw new Saml2Exception("Invalid SAML response destination: " + samlResponse.getDestination());
+			throw authException(INVALID_DESTINATION, "Invalid SAML response destination: " + samlResponse.getDestination());
 		}
 
 		final String issuer = samlResponse.getIssuer().getValue();
 		if (logger.isDebugEnabled()) {
-			logger.debug("Processing SAML response from " + issuer);
-		}
-		if (token == null) {
-			throw new Saml2Exception(format("SAML 2 Provider for %s was not found.", issuer));
+			logger.debug("Validating SAML response from " + issuer);
 		}
+
+		Saml2AuthenticationException lastValidationError = null;
+
 		boolean responseSigned = hasValidSignature(samlResponse, token);
 		for (Assertion a : samlResponse.getAssertions()) {
 			if (logger.isDebugEnabled()) {
 				logger.debug("Checking plain assertion validity " + a);
 			}
-			if (isValidAssertion(recipient, a, token, !responseSigned)) {
-				if (logger.isDebugEnabled()) {
-					logger.debug("Found valid assertion. Skipping potential others.");
-				}
+			try {
+				validateAssertion(recipient, a, token, !responseSigned);
 				return a;
+			} catch (Saml2AuthenticationException e) {
+				lastValidationError = e;
 			}
 		}
 		for (EncryptedAssertion ea : samlResponse.getEncryptedAssertions()) {
 			if (logger.isDebugEnabled()) {
 				logger.debug("Checking encrypted assertion validity " + ea);
 			}
-
-			Assertion a = decrypt(token, ea);
-			if (isValidAssertion(recipient, a, token, false)) {
-				if (logger.isDebugEnabled()) {
-					logger.debug("Found valid encrypted assertion. Skipping potential others.");
-				}
+			try {
+				Assertion a = decrypt(token, ea);
+				validateAssertion(recipient, a, token, false);
 				return a;
+			} catch (Saml2AuthenticationException e) {
+				lastValidationError = e;
 			}
 		}
-		throw new InsufficientAuthenticationException("Unable to find a valid assertion");
+		if (lastValidationError != null) {
+			throw lastValidationError;
+		}
+		else {
+			throw authException(MALFORMED_RESPONSE_DATA, "No assertions found in response.");
+		}
 	}
 
-	private boolean hasValidSignature(SignableSAMLObject samlResponse, Saml2AuthenticationToken token) {
-		if (!samlResponse.isSigned()) {
+	private boolean hasValidSignature(SignableSAMLObject samlObject, Saml2AuthenticationToken token) {
+		if (!samlObject.isSigned()) {
+			if (logger.isDebugEnabled()) {
+				logger.debug("SAML object is not signed, no signatures found");
+			}
 			return false;
 		}
 
-		final List<X509Certificate> verificationKeys = getVerificationKeys(token);
+		final List<X509Certificate> verificationKeys = getVerificationCertificates(token);
```

Comment:
To align with the rest of the codebase, please leave the `final` keyword off of local variables.